xmtp_mls/README.md

```diff
@@ -0,0 +1,143 @@
+# XMTP MLS
+
+## Database Schema
+
+Foreign key constraints and indexes omitted for simplicity.
+
+```sql
+-- Would be a table that only holds a single identity (your own)
+CREATE TABLE identities {
+    id TEXT PRIMARY KEY NOT NULL, -- The installation_id of the identity
+    signature_keypair BLOB NOT NULL,
+    "credential" BLOB NOT NULL,
+    created_at_ns BIGINT NOT NULL,
+    wallet_address TEXT NOT NULL,
+}
+
+-- Minimum viable group schema
+CREATE TABLE groups {
+    id BLOB PRIMARY KEY NOT NULL, -- Random ID generated by group creator
+    created_at_ns BIGINT NOT NULL, -- Based on the timestamp of the welcome message
+    membership_state INT NOT NULL, -- Enum of the GROUP_MEMBERSHIP_STATE listed below
+}
+
+-- Successfully processed messages meant to be returned to the user
+CREATE TABLE group_messages {
+    id BLOB PRIMARY KEY NOT NULL -- Derived via SHA256(CONCAT(decrypted_message_bytes, conversation_id, timestamp))
+    group_id BLOB NOT NULL,
+    decrypted_message_bytes BLOB NOT NULL, -- Message contents after decryption
+    sent_at_ns BIGINT NOT NULL, -- Based on the timestamp of the message
+    sender_installation_id TEXT NOT NULL,
+    sender_wallet_address TEXT NOT NULL, -- Could remove this if we added a table mapping installation_ids to wallet addresses
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id BLOB NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash BLOB -- The hash of the encrypted, concrete, form of the message if it was published.
+}
+
+CREATE TABLE outbound_welcome_messages {
+    id BLOB PRIMARY KEY NOT NULL, -- Derived via SHA256(CONCAT(group_id, welcome_message, ratchet_tree, installation_id))
+    state INT NOT NULL, -- OUTBOUND_WELCOME_STATE
+    installation_id BLOB PRIMARY KEY NOT NULL,
+    commit_hash BLOB NOT NULL, -- The hash of the commit which created this
```

Comment:
Does this correspond to message_hash in group_intents? Same column name would make that clearer

Reply:
Good call out. I'll rename.